x-pack/plugins/fleet/public/search_provider.ts

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type { CoreSetup, CoreStart } from 'src/core/public';
+
+import type { Observable } from 'rxjs';
+import { from, of, combineLatest } from 'rxjs';
+import { map, shareReplay, takeUntil } from 'rxjs/operators';
+
+import type {
+  GlobalSearchResultProvider,
+  GlobalSearchProviderResult,
+} from '../../global_search/public';
+
+import { INTEGRATIONS_PLUGIN_ID } from '../common';
+
+import { sendGetPackages } from './hooks';
+import type { GetPackagesResponse } from './types';
+import { pagePathGetters } from './constants';
+
+const packageType = 'package';
+
+export const createPackageSearchProvider = (core: CoreSetup): GlobalSearchResultProvider => {
+  const coreStart$ = from(core.getStartServices()).pipe(
+    map(([coreStart]) => coreStart),
+    shareReplay(1)
+  );
+
+  let packages$: undefined | Observable<GetPackagesResponse['response']>;
+
+  const getPackages$ = () => {
+    if (!packages$) {
```

Comment:
It might make more sense to invert the conditional here and return early if we have package data memorized, just to bring the bulk of this function body out of the `if` block.

```ts
if (packages$) {
  return packages$;
}

packages$ = from(...
```

Reply:
In this case I am inclined to go with the solution that provides a single `return` but I agree that the body of the if is a bit verbose. I've refactored to:

```ts
const createPackages$ = () =>
  from(sendGetPackages()).pipe(
    map(({ error, data }) => {
      if (error) {
        throw error;
      }
      return data?.response ?? [];
    }),
    shareReplay(1)
  )
...
  const getPackages$ = () => {
    if (!packages$) {
      packages$ = createPackages$();
    }
    return packages$;
  };
```

Let me know what you think!

Reply:
Looks great! 🚢